### PR TITLE
Generate syft.ts in cli init

### PR DESCRIPTION
**Pull request by finnsyft**
TASK: [SYF-65](https://linear.app/syft/issue/SYF-65/)

This change adds a syft.ts file into a src folder when running the syft init cli command. It creates an enum of metrics dependencies with corresponding supported metrics provider package names. This is then used to parse a project's package.json file for providers and adds those to the generated syft.ts file.

### CODE REVIEW

**Comment by finnsyft** [comment]
Initial commit WIP. We need to discuss whether we want to have this in the init command or have a separate cli command. Unit tests will be added after we confer.

File: packages/client/src/plugins/index.ts [hunk 1]
```diff
@@ -1,3 +1,18 @@
+/**
+ * Supported Metrics Providers. We can revisit modeling this in the future.
+ */
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-nocheck
+export enum MetricProvider {
+  Amplitude = 'amplitude-js' as string,
```
Comment (naagas-syft) [inline packages/client/src/plugins/index.ts:9]: why is "as string" required ?

File: packages/cli/src/commands/init.ts [hunk 2]
```diff
@@ -74,11 +84,26 @@ function initalizeSchemaFolder(folder: string, force: boolean): boolean {
...
-function generateSchemasFrom(ast: AST, folder: string): void {
+async function getMetricPluginList(dir: string): Promise<string[]> {
+  const filepath = path.join(dir, 'package.json');
+  const deps = JSON.parse(fs.readFileSync(filepath, 'utf8')).dependencies;
+
+  const providers: string[] = [];
+  if (deps != null) {
+    for (const providerPackage of Object.values(MetricProvider)) {
+      if (providerPackage in deps) {
+        providers.push(`${MetricProvider[providerPackage]}` + 'Plugin');
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts]: nit: i'd explicitly maintain the mapping between a provider and the plugin class name. that way i can find usages of the plugin via a text search. may be write a comment in the index file saying that the id and plugin name need to use the same ?

File: packages/cli/src/commands/init.ts [hunk 3]
```diff
@@ -74,11 +84,26 @@ function initalizeSchemaFolder(folder: string, force: boolean): boolean {
...
+  }
+  return providers;
+}
+
+async function generateSchemasFrom(ast: AST, folder: string): Promise<void> {
   logVerbose(`Output dir: ${folder}`);
   // generate basic assets.
   const lintRules = path.join(folder, 'lint/rules');
-  if (!fs.existsSync(lintRules)) {
+  if (fs.existsSync(lintRules) != null) {
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts]: existsSync returns boolean right ?

File: packages/cli/src/commands/init.ts [hunk 4]
```diff
@@ -29,6 +30,15 @@ export interface Params {
 
 export const command = 'init [platform] [product]';
 export const desc = 'Initialize Syft project. Generates Event models.';
+
+const syftFileContent =
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts]: nit: may be move this to `https://github.com/syftdata/syft/tree/main/packages/cli/assets`

File: packages/cli/src/commands/init.ts [hunk 5]
```diff
@@ -120,6 +146,33 @@ function generateSchemasFrom(ast: AST, folder: string): void {
...
+        metricsPluginList
+          .map((input: string) => `    new ${input}()`)
+          .join(',\n') +
+        '\n  ';
+
+  const content = syftFileContent
+    .replace(/\{ IMPORT }/, imports)
+    .replace(/\{ INSTANTIATION }/, instantiations);
+  const syftSrcFile = project.createSourceFile(
+    path.join(folder, '/src/syft.ts'),
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts]: looks like you are creating the file under "@syftdata/client" project. `folder` point to that i believe. did you test the location of syft.ts file after running the cli ?

**Review by naagas-syft — COMMENTED** [review]
left comments.

File: packages/cli/src/commands/init.ts [hunk 6]
```diff
@@ -123,6 +124,82 @@ function generateSchemasFrom(ast: AST, folder: string): void {
...
+  if (metricPlugins.length > 0) {
+    logVerbose(`Generating syft.ts with ${metricPlugins.join(', ')} plugin(s)`);
+  } else {
+    logVerbose('No compatible plugins found. Not generating syft.ts');
+    return;
+  }
+
+  const destinationPath = path.join('src', 'syft.ts');
+  if (fs.existsSync(destinationPath) && !force) {
+    logVerbose(`${destinationPath} already exists. Use --force to overwrite.`);
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts]: logInfo ?

File: packages/cli/src/commands/init.ts [hunk 7]
```diff
@@ -123,6 +124,82 @@ function generateSchemasFrom(ast: AST, folder: string): void {
...
+    `Parsed package.json file and found ${providers.length} compatible plugins.`
+  );
+  return providers;
+}
+
+function generateSyftTS(metricPlugins: string[], force: boolean): void {
+  if (metricPlugins.length > 0) {
+    logVerbose(`Generating syft.ts with ${metricPlugins.join(', ')} plugin(s)`);
+  } else {
+    logVerbose('No compatible plugins found. Not generating syft.ts');
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts]: logInfo ?

File: packages/cli/src/commands/init.ts [hunk 8]
```diff
@@ -123,6 +124,82 @@ function generateSchemasFrom(ast: AST, folder: string): void {
...
+
+  const imports =
+    metricPlugins.length === 1
+      ? ` ${metricPlugins[0]} `
+      : '\n  ' + metricPlugins.join(',\n  ') + '\n';
+  const instantiations =
+    metricPlugins.length === 1
+      ? `new ${metricPlugins[0]}()`
+      : '\n' +
+        metricPlugins.map((input: string) => `    new ${input}()`).join(',\n') +
```
Comment (naagas-syft) [inline packages/cli/src/commands/init.ts:168]: nit: wondering if we can trigger an eslint to re-format the code. generating 4-spaces may not be the right thing.  may be add a TODO note ?

File: packages/client/package.json [hunk 9]
```diff
@@ -68,4 +68,4 @@
     "zod": "^3.20.5"
   },
   "gitHead": "88c566d72818936b193d860ced18df01b84bf140"
-}
+}
```
Comment (naagas-syft) [inline packages/client/package.json]: unnecessary change.

File: packages/cli/tests/test_package/srcWithSyft/syft.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+//NOT CHANGED
```
Comment (naagas-syft) [inline packages/cli/tests/test_package/srcWithSyft/syft.ts:1]: hmm. is this an output of a test ?

**Review by naagas-syft — APPROVED** [review]
looks good.

File: packages/cli/tests/test_package/srcWithSyft/syft.ts (hunk 10 above)
Comment (finnsyft) [reply packages/cli/tests/test_package/srcWithSyft/syft.ts:1]: it's just there to support 'init does not create syft file if it already exists'. since everything is testing the file system, this was the most consistent way to support the test case

File: packages/cli/src/commands/init.ts (hunk 8 above)
Comment (finnsyft) [reply packages/cli/src/commands/init.ts:168]: https://linear.app/syft/issue/SYF-71/use-eslint-on-to-reformat-syftts-on-init
